### PR TITLE
NLP Java Classes

### PR DESCRIPTION
**Pull request by yimingnzhao**
Classes for Java objects used in the NLP pipeline. Note that these classes must be constructed and called by a servlet.

### CODE REVIEW

File: NLP_Files/NaturalLanguagePostprocessor.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        for (String entity : entities) {
+            if (!entitiesMap.containsKey(entity)) {
+                entitiesMap.put(entity, new ArrayList<TimeRange>());
+            } 
+            entitiesMap.get(entity).add(currentTimeRange);
+        }
+    }
+
+    /**
+     * Gets the entites mapping that is currently built
```
Comment (RickyAvina) [inline NLP_Files/NaturalLanguagePostprocessor.java]: entities

File: NLP_Files/NaturalLanguagePostprocessor.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ArrayList;
+import com.google.sps.data.TimeRange;
+
+/** Postprocessing for outputs of the NLP API that maps a keyphrase to all mention times */
+public class NaturalLanguagePostprocessor {
+    
+    private Map<String, List<TimeRange>> entitiesMap;
```
Comment (RickyAvina) [inline NLP_Files/NaturalLanguagePostprocessor.java]: would be nice to document the function of this map

File: NLP_Files/NaturalLanguagePreprocessor.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+package com.google.sps.data;
+
+import java.util.List;
+import java.util.ArrayList;
+import com.google.sps.data.TimeRangedText;
+
+/** Preprocesses captions by concatenating TimeRangedTexts to reach a time threshold */
+public class NaturalLanguagePreprocessor {
+
+    private static final long DEFAULT_TIME_THRESHOLD = 20;
```
Comment (RickyAvina) [inline NLP_Files/NaturalLanguagePreprocessor.java]: A comment here would also be good explaining this constant.

File: NLP_Files/NaturalLanguageProcessor.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+     * Gets a list of entities using Google's NLP API
+     * @param text The string that contains the text to pull entities from
+     * @return The list of entities
+     */
+    public List<String> getEntities(String text) {
+        Set<String> entities = new HashSet<String>();
+        int tokens = text.split(WORD_DELIMITER).length;
+        boolean hasAcademicCategory = false;
+
+        // Calls the NLP API, but returns null if exceptions occur
```
Comment (RickyAvina) [inline NLP_Files/NaturalLanguageProcessor.java]: Where can errors be thrown here?

File: NLP_Files/NaturalLanguageProcessor.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    continue;
+                }
+                String target = category.getName().split(CATEGORY_DELIMITER)[CATEGORY_INDEX];
+
+                if (Arrays.stream(ACADEMIC_CATEGORIES).anyMatch(x -> target.equals(x))) {
+                    return true;
+                }
+            }
+        } catch (Exception e) {
+            return false;
```
Comment (RickyAvina) [inline NLP_Files/NaturalLanguageProcessor.java]: Where would this error arise and is this the behavior you want?

**Comment by yimingnzhao** [comment]
Code merged with the 'merged-youtube-nlp' branch

**Pull request closed without merging**